### PR TITLE
Move webhook to a Deployment

### PR DESCRIPTION
**Pull request by qinqon**
**Is this a BUG FIX or a FEATURE ?**:
/kind bug

**What this PR does / why we need it**:
This PR moves the webhook to a unprivileged `Deployment` without `hostNetwork: true`, this prevent port collision since the webhook port will be open at pod's netns.

It also runs `nmstate-handler` only at workers this is the common scenario and remove the WEBHOOK_PORT env var since it's no longer needed and remove complexity a little.

**Special notes for your reviewer**:
We run with two replicas so cert-manager using LeaderElection is being tested already.

**Release note**:

```release-note
Webhook runs now as unprivileged `Deployment`
action requiered: Wait for webhook Deployment Ready status too after deploy/upgrade.
```

### CODE REVIEW

File: pkg/webhook/nodenetworkconfigurationpolicy/server.go [hunk 1]
```diff
@@ -38,7 +21,6 @@ func Add(mgr manager.Manager) error {
 	// 2.- Since we have delete the condition the status-mutate webhook get called and
 	//     there we set conditions to Unknown this final result will be updated.
 	server := webhookserver.New(mgr, webhookName, certificate.MutatingWebhook,
-		webhookserver.WithPort(webhookPort),
```
Comment (phoracek) [inline pkg/webhook/nodenetworkconfigurationpolicy/server.go:41]: We still have targetPort in service. How does it work now without explicit port being specified?

File: pkg/helper/client.go [hunk 2]
```diff
@@ -37,6 +37,11 @@ var (
 )
 
 func init() {
+	_, runWebhookServer := os.LookupEnv("RUN_WEBHOOK_SERVER")
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/helper/client.go:40]: Yep

File: pkg/webhook/nodenetworkconfigurationpolicy/server.go (hunk 1 above)
Comment (qinqon) [reply pkg/webhook/nodenetworkconfigurationpolicy/server.go:41]: It will use kube-admission-webhook default port https://github.com/qinqon/kube-admission-webhook/blob/master/pkg/webhook/server/server.go#L35

Comment (phoracek) [reply pkg/webhook/nodenetworkconfigurationpolicy/server.go:41]: Then you gotta fix the manifest, `Service` still has `      targetPort: 54874`

File: deploy/operator.yaml [hunk 3]
```diff
@@ -223,3 +191,14 @@ webhooks:
...
         apiVersions: ["v1alpha1"]
         resources: ["nodenetworkconfigurationpolicies", "nodenetworkconfigurationpolicies/status"]
+---
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  namespace: {{ .HandlerNamespace }}
+  name: {{template "handlerPrefix" .}}nmstate-webhook
+spec:
+  minAvailable: "90%"
```
Comment (phoracek) [inline deploy/operator.yaml:201]: We have two replicas, 90% would mean we need 2/2. I think we should set it so it requires at least one instance to be up and running.

File: deploy/operator.yaml [hunk 4]
```diff
@@ -46,69 +45,38 @@ spec:
...
         app: kubernetes-nmstate
-        name: {{template "handlerPrefix" .}}nmstate-handler-worker
+        name: {{template "handlerPrefix" .}}nmstate-handler
     spec:
       # Needed to force vlan filtering config with iproute commands until
       # future nmstate/NM is in place.
       # https://github.com/nmstate/nmstate/pull/440
       hostNetwork: true
       serviceAccountName: {{template "handlerPrefix" .}}nmstate-handler
-      affinity:
```
Comment (phoracek) [inline deploy/operator.yaml:103]: We need to add the `toleration` we had in the master.

File: deploy/operator.yaml [hunk 5]
```diff
@@ -1,24 +1,23 @@
...
   selector:
     matchLabels:
-      name: {{template "handlerPrefix" .}}nmstate-handler
+      name: {{template "handlerPrefix" .}}nmstate-webhook
   template:
     metadata:
       labels:
         app: kubernetes-nmstate
-        name: {{template "handlerPrefix" .}}nmstate-handler
+        name: {{template "handlerPrefix" .}}nmstate-webhook
```
Comment (phoracek) [inline deploy/operator.yaml:19]: * please rename the container from the pod as well. it is still "handler"
* I believe we don't need to set WATCH_NAMESPACE since we don't have any controller, do we? are we using POD_NAME?

File: pkg/webhook/nodenetworkconfigurationpolicy/server.go (hunk 1 above)
Comment (phoracek) [reply pkg/webhook/nodenetworkconfigurationpolicy/server.go:41]: https://github.com/nmstate/kubernetes-nmstate/pull/497/files#diff-5e5f804a8ee2fb8c81711b728eb5a21dR150 still applies

File: deploy/operator.yaml (hunk 5 above)
Comment (qinqon) [reply deploy/operator.yaml:19]: Let's remove all we can to reduce stuff and see if it works.

Comment (qinqon) [reply deploy/operator.yaml:19]: WHACH_NAMESPACE is needed

File: deploy/handler/operator.yaml [hunk 6]
```diff
@@ -39,78 +40,49 @@ spec:
...
-      - name: dbus-socket
-        hostPath:
-          path: /run/dbus/system_bus_socket
-          type: Socket
-      - name: nmstate-lock
-        hostPath:
-          path: /var/k8s_nmstate
 ---
 apiVersion: apps/v1
 kind: DaemonSet
```
Comment (yboaron) [inline deploy/handler/operator.yaml:57]: Do we have a single daemonset? I can't find the workers Ds.

**Comment by bcrochet** [comment]
Please rebase.

**Comment by qinqon** [comment]
@bcrochet done, but I have to fix something there there are some race conditions at webhooks and controller-rutime cache and also I have to fix the cert-manager.

**Comment by qinqon** [comment]
Manual test on openshift 4.5 works fine with a linux bridge without ports (just simple use case)

```bash
[ellorent@localhost kubernetes-nmstate]$ oc get nncp
NAME   STATUS
lbr1   SuccessfullyConfigured
[ellorent@localhost kubernetes-nmstate]$ oc get nnce
NAME                                    STATUS
lbednar-ocp45-8s8t2-master-0.lbr1       SuccessfullyConfigured
lbednar-ocp45-8s8t2-master-1.lbr1       SuccessfullyConfigured
lbednar-ocp45-8s8t2-master-2.lbr1       SuccessfullyConfigured
lbednar-ocp45-8s8t2-worker-268bg.lbr1   SuccessfullyConfigured
lbednar-ocp45-8s8t2-worker-cwnhs.lbr1   SuccessfullyConfigured
lbednar-ocp45-8s8t2-worker-fr6wm.lbr1   SuccessfullyConfigured
```

File: cmd/manager/main.go [hunk 7]
```diff
@@ -124,18 +135,18 @@ func main() {
...
-	if err := controller.AddToManager(mgr); err != nil {
-		log.Error(err, "Cannot initialize controller")
-		os.Exit(1)
-	}
-
-	// Setup webhook on master only
-	if _, runWebhookServer := os.LookupEnv("RUN_WEBHOOK_SERVER"); runWebhookServer {
+	// Runs only webhook controllers if it's specified
+	if environment.IsWebhook() {
 		if err := webhook.AddToManager(mgr); err != nil {
```
Comment (yboaron) [inline cmd/manager/main.go:140]: Any special reason not to add controllers in the same approach for all controllers? 
I mean for operator and handlers, we call all controllers, and each controller decides if it should be registered or not, and for webhook case, we only Add the relevant controller here.

File: deploy/handler/operator.yaml [hunk 8]
```diff
@@ -1,37 +1,35 @@
 {{define "handlerPrefix"}}{{with $prefix := .HandlerPrefix}}{{$prefix | printf "%s-"}}{{end -}}{{end}}
 ---
 apiVersion: apps/v1
-kind: DaemonSet
+kind: Deployment
```
Comment (yboaron) [inline deploy/handler/operator.yaml:4]: Q: We want the webhook pods to run on Masters only because of performance (k8s-api run on masters), right?

File: deploy/handler/operator.yaml [hunk 9]
```diff
@@ -42,81 +40,74 @@ spec:
...
-                  - key: node-role.kubernetes.io/master
-                    operator: NotIn
-                    values:
-                      - ""
       nodeSelector:
         beta.kubernetes.io/arch: amd64
       {{- range $key, $value := .HandlerNodeSelector }}
       {{"  "}}{{- $key }}: {{  $value }}
       {{- end}}
+      tolerations:
```
Comment (yboaron) [inline deploy/handler/operator.yaml:107]: I can't see the master handlers DS, do we want to limit the handlers to run only on worker nodes?

File: pkg/controller/nmstate/nmstate_controller_test.go [hunk 10]
```diff
@@ -109,24 +109,32 @@ var _ = Describe("NMState controller reconcile", func() {
...
-					Expect(ds.Spec.Template.Spec.NodeSelector).To(HaveKeyWithValue(k, v))
-				}
+			handlerKey := types.NamespacedName{Namespace: handlerNamespace, Name: handlerPrefix + "-nmstate-handler"}
+			err := cl.Get(context.TODO(), handlerKey, ds)
+			Expect(err).ToNot(HaveOccurred())
+			for k, v := range handlerNodeSelector {
+				Expect(ds.Spec.Template.Spec.NodeSelector).To(HaveKeyWithValue(k, v))
+			}
+		})
+		It("should NOT add NodeSelector to webhook deployment", func() {
```
Comment (yboaron) [inline pkg/controller/nmstate/nmstate_controller_test.go:131]: Any special reason to add this test?  you didn't include the  '.HandlerNodeSelector' in webhook deployment manifest?

File: deploy/handler/operator.yaml [hunk 11]
```diff
@@ -42,81 +40,74 @@ spec:
...
-                fieldRef:
-                  fieldPath: spec.nodeName
-            - name: INTERFACES_FILTER
-              valueFrom:
-                configMapKeyRef:
-                  name: {{template "handlerPrefix" .}}nmstate-config
-                  key: interfaces_filter
-            - name: WEBHOOK_PORT
-              value: "54874"
+              value: "{{template "handlerPrefix" .}}nmstate"
```
Comment (yboaron) [inline deploy/handler/operator.yaml:50]: Why does the webhook deployment need to know the Operator name?

File: cmd/manager/main.go [hunk 12]
```diff
@@ -81,9 +81,10 @@ func main() {
 
 	printVersion()
 
-	if !environment.IsOperator() {
-		// Take exclusive instance lock, we need that to make sure that
-		// we don't have more than one working instance of k8s-nmstate
+	// Lock only for handler we can run old and new version of
```
Comment (yboaron) [inline cmd/manager/main.go:84]: nit, add coma  ( 'Lock only for handler, bla bla )

File: deploy/handler/operator.yaml (hunk 9 above)
Comment (yboaron) [reply deploy/handler/operator.yaml:107]: I just read in the PR description that we removing the handlers to run only on workers by default. 
What the user should do if he wants to include also the masters?
If we want to change for example the k8s control plane network using nmstate ( for example run it as tagged traffic), I guess we need nmstate to run also on masters for that case

File: cmd/manager/main.go (hunk 7 above)
Comment (qinqon) [reply cmd/manager/main.go:140]: webhook are not controllers they are different concept at controler-runtime so they have different line.

File: deploy/handler/operator.yaml (hunk 8 above)
Comment (qinqon) [reply deploy/handler/operator.yaml:4]: Is more about stability than performance, users will configure networking usually at workers so workers are going to be more instable alas breaking the webhook if we run it there also if you think about it the webhook is kind of knmstate control-plane so it belongs to masters.

File: deploy/handler/operator.yaml (hunk 9 above)
Comment (qinqon) [reply deploy/handler/operator.yaml:107]: This PR has one type of handler daemonset that runs at master and workers, if we want to change just masters we can put in place a profile with nodeSelector `kubernetes.io/role/master=""`

File: pkg/controller/nmstate/nmstate_controller_test.go (hunk 10 above)
Comment (qinqon) [reply pkg/controller/nmstate/nmstate_controller_test.go:131]: The operator's CR NodeSelector is only for handlers (we want to skip nodes that does not support nmstate) webhook is transparent for that since is just ensuring the nncp status is correctly reset.

File: deploy/handler/operator.yaml (hunk 11 above)
Comment (qinqon) [reply deploy/handler/operator.yaml:50]: Manager was failing operator-sdk need it (I don't remeber, also we need it to point to MutatingWebhookConfiguration)

File: test/runner/pod.go [hunk 13]
```diff
@@ -10,23 +10,23 @@ import (
...
+func runAtPods(pods []string, arguments ...string) {
+	for _, pod := range pods {
+		exec := []string{"exec", "-n", framework.Global.Namespace, pod, "--"}
 		execArguments := append(exec, arguments...)
 		_, err := cmd.Kubectl(execArguments...)
-		ExpectWithOffset(1, err).ToNot(HaveOccurred())
+		ExpectWithOffset(2, err).ToNot(HaveOccurred())
 	}
 }
+func RunAtHandlerPods(arguments ...string) {
```
Comment (phoracek) [inline test/runner/pod.go:28]: Please add an empty line above

File: cluster/sync.sh [hunk 14]
```diff
@@ -41,16 +32,25 @@ function consistently {
     done
 }
 
-function isHandlerOk {
-    desiredNumberScheduled=$(getDesiredScheduledHandlers $1)
-    numberAvailable=$(getAvailableHandlers $1)
+function isDaemonSetOk {
+
```
Comment (phoracek) [inline cluster/sync.sh:36]: Please drop the line

File: cluster/sync.sh [hunk 15]
```diff
@@ -41,16 +32,25 @@ function consistently {
...
+    desiredNumberScheduled=$($kubectl get daemonset -n $1 -l $2 -o=jsonpath='{..status.desiredNumberScheduled}')
```
Comment (phoracek) [inline cluster/sync.sh:37]: Why do we have two dots in the jsonpath?

File: cmd/manager/main.go [hunk 16]
```diff
@@ -81,9 +81,10 @@ func main() {
 
 	printVersion()
 
-	if !environment.IsOperator() {
-		// Take exclusive instance lock, we need that to make sure that
-		// we don't have more than one working instance of k8s-nmstate
+	// Lock only for handler, we can run old and new version of
+	// webhook without problems, policy status will be updated
+	// by multiple instances.
+	if environment.IsHandler() {
```
Comment (phoracek) [inline cmd/manager/main.go:87]: The fact that we need this kind of `if` several times in the main function suggests we should maybe create a function mainWebhook, mainHandler, mainOperator. They would share some logic, but the result would be way more readable. If somebody wants to follow execution of a webhook, they just jump to that function.

File: deploy/handler/operator.yaml [hunk 17]
```diff
@@ -1,37 +1,35 @@
 {{define "handlerPrefix"}}{{with $prefix := .HandlerPrefix}}{{$prefix | printf "%s-"}}{{end -}}{{end}}
 ---
 apiVersion: apps/v1
-kind: DaemonSet
+kind: Deployment
 metadata:
-  name: {{template "handlerPrefix" .}}nmstate-handler
```
Comment (phoracek) [inline deploy/handler/operator.yaml:6]: We will need to add removal code for this to CNAO. Like the one we have for multus.

File: test/e2e/handler/prepare.go [hunk 18]
```diff
@@ -59,3 +56,24 @@ func waitForDaemonSets(t *testing.T, kubeclient kubernetes.Interface, namespace
 	})
 	return err
 }
+
+func waitForDeployments(t *testing.T, kubeclient kubernetes.Interface, namespace string, retryInterval, timeout time.Duration) error {
+	if framework.Global.LocalOperator {
+		return nil
+	}
+	err := wait.PollImmediate(retryInterval, timeout, func() (done bool, err error) {
```
Comment (phoracek) [inline test/e2e/handler/prepare.go:64]: Let's make this breathe, please insert an empty line above this one.

File: test/e2e/handler/prepare.go [hunk 19]
```diff
@@ -59,3 +56,24 @@ func waitForDaemonSets(t *testing.T, kubeclient kubernetes.Interface, namespace
...
+		filterByApp := metav1.ListOptions{LabelSelector: "app=kubernetes-nmstate"}
+		deployments, err := kubeclient.AppsV1().Deployments(namespace).List(filterByApp)
+		if err != nil {
+			return true, errors.Wrapf(err, "failed retrieving daemon sets for namespace %s", namespace)
+		}
+		for _, deployment := range deployments.Items {
```
Comment (phoracek) [inline test/e2e/handler/prepare.go:70]: ditto

File: test/e2e/handler/prepare.go [hunk 20]
```diff
@@ -59,3 +56,24 @@ func waitForDaemonSets(t *testing.T, kubeclient kubernetes.Interface, namespace
...
+			By(fmt.Sprintf("Checking deployment %s", deployment.Name))
+			if int(deployment.Status.AvailableReplicas) >= 2 {
+				return true, nil
+			}
+		}
+		return true, nil
+	})
+	return err
```
Comment (phoracek) [inline test/e2e/handler/prepare.go:78]: ditto

File: test/e2e/handler/prepare.go [hunk 21]
```diff
@@ -59,3 +56,24 @@ func waitForDaemonSets(t *testing.T, kubeclient kubernetes.Interface, namespace
...
+		if err != nil {
+			return true, errors.Wrapf(err, "failed retrieving daemon sets for namespace %s", namespace)
+		}
+		for _, deployment := range deployments.Items {
+			By(fmt.Sprintf("Checking deployment %s", deployment.Name))
+			if int(deployment.Status.AvailableReplicas) >= 2 {
+				return true, nil
+			}
+		}
+		return true, nil
```
Comment (phoracek) [inline test/e2e/handler/prepare.go:76]: ditto

File: pkg/environment/environment.go [hunk 22]
```diff
@@ -0,0 +1,25 @@
+package environment
```
Comment (phoracek) [inline pkg/environment/environment.go:1]: I won't block this PR on this, but this is wrong. It keeps functions in a single module grouped by the function they use, not by their "unit". That makes this into a junkyard of env variable handling.

This points at the issue that we are coupling everything together here. We should have completely separated lane for every of these modes (operator, handler, webhook). We should read the parameter (IsOperator) in main and just run the right component. Just ranting, I think we will need to make some order in the code-base, sooner or later.

File: pkg/controller/nmstate/nmstate_controller_test.go [hunk 23]
```diff
@@ -26,7 +26,7 @@ var _ = Describe("NMState controller reconcile", func() {
 		cl                  client.Client
 		reconciler          ReconcileNMState
 		existingNMStateName = "nmstate"
-		dsNodeSelector      = map[string]string{"selector_1": "value_1", "selector_2": "value_2"}
+		handlerNodeSelector = map[string]string{"selector_1": "value_1", "selector_2": "value_2"}
```
Comment (phoracek) [inline pkg/controller/nmstate/nmstate_controller_test.go:29]: so many unrelated changes making this harder to review and spot real issues, tsk tsk tsk.

File: deploy/handler/role.yaml [hunk 24]
```diff
@@ -76,10 +76,9 @@ rules:
   - pods
   - nodes
   - namespaces
+  - secrets
   verbs:
-  - get
-  - list
-  - watch
+  - '*'
```
Comment (phoracek) [inline deploy/handler/role.yaml:81]: We only need to create secrets, rights? Please add a new item in this list specifically for secrets and verbs they need.

File: cluster/sync.sh (hunk 15 above)
Comment (qinqon) [reply cluster/sync.sh:37]: To simplify expression and just read the first occurrence of that field now we have just one DaemonSet

File: cmd/manager/main.go (hunk 16 above)
Comment (qinqon) [reply cmd/manager/main.go:87]: Will do a follow up PR.

File: pkg/environment/environment.go (hunk 22 above)
Comment (qinqon) [reply pkg/environment/environment.go:1]: At the follow-up PR we will refactor this.

File: deploy/handler/operator.yaml (hunk 17 above)
Comment (qinqon) [reply deploy/handler/operator.yaml:6]: noted

File: deploy/handler/role.yaml [hunk 25]
```diff
@@ -77,9 +84,15 @@ rules:
   - nodes
   - namespaces
   verbs:
+  - '*'
```
Comment (phoracek) [inline deploy/handler/role.yaml:87]: Do we need to keep this changed now when we moved secrets out?
